opensearch/aryn-opensearch.sh

```diff
@@ -0,0 +1,427 @@
+               -o "${file}" \
+               --data @- <<END || die "Error registering GPT model"
+{
+  "name": "openAI-gpt-3.5-turbo",
+  "function_name": "remote",
+  "description": "gpt model",
+  "connector_id": "${OPENAI_CONN_ID}",
+  "model_group_id": "${MODEL_GROUP_ID}"
+}
+END
+
+    local id="$(jq -r '.task_id' "${file}")"
+    [[ -z "${id}" ]] && die "No OpenAI task ID from registering model"
+
+    OPENAI_TASK_ID="${id}"
+    echo "OPENAI_TASK_ID=${OPENAI_TASK_ID}" >>"${PERSISTENT_ENV_TMP}"
+}
+
+deploy_model() {
+    local model_id="$1" # if empty will be derived from task
+    local task_id="$2"
+    local name="$3"
+    [[ -z "${task_id}" ]] && die "ERROR missing task_id"
+    [[ -z "${name}" ]] && die "ERROR missing name"
+
+    if [[ -z "${model_id}" ]]; then
+        wait_task "${task_id}" "$name model to become ready"
+        model_id="${GET_TASK_MODEL_ID}"
+        echo "Fetched $name model id $model_id for task_id $task_id"
+    else
+        echo "Using existing model id ${model_id}"
+    fi
+    MODEL_ID="${model_id}"
+    DEPLOY_MODEL_LOG_FILE="${ARYN_STATUSDIR}/curl.deploy_model_task.${name}"
+    wait_or_die deploy_model_try "deploy of model ${MODEL_ID}, task ${task_id}, name ${name}" 60
+    wait_task "${DEPLOY_MODEL_DEP_TASK_ID}" "$name model to deploy"
+    # TODO: eric - unclear if we're waiting for the deploy to finish; log message about task done shows up in opensearch after
+    # restart logic finishes
+}
+
+wait_task() {
+    local task_id="$1"
+    local msg="$2"
+
+    [[ -z "${task_id}" ]] && die "missing task_id"
+    [[ -z "${msg}" ]] && die "missing msg"
+
+    GET_TASK_ID="${task_id}"
+    wait_or_die get_task "${msg}" 60
+}
+
+get_task() {
+    local file="${ARYN_STATUSDIR}/curl.get_task"
+    rm -f "${file}" 2>/dev/null
+    _curl "${BASE_URL}/_plugins/_ml/tasks/${GET_TASK_ID}" -o "${file}"
+
+    [[ -r "${file}" ]] || return 1
+    local model="$(jq -r '.model_id' "${file}")"
+    [[ -z "${model}" || "${model}" = 'null' ]] && return 1
+
+    GET_TASK_MODEL_ID="${model}"
+    return 0
+}
+
+deploy_model_try() {
+    _curl -X POST "${BASE_URL}/_plugins/_ml/models/${MODEL_ID}/_deploy" \
+          -o "${DEPLOY_MODEL_LOG_FILE}"
+    local dep_task_id="$(jq -r '.task_id' "${DEPLOY_MODEL_LOG_FILE}")"
+    [[ -z "${dep_task_id}" || "${dep_task_id}" == null ]] \
+        && return 1
+    DEPLOY_MODEL_DEP_TASK_ID="${dep_task_id}"
+}
+
+sp_create_rag_pipeline() {
+    _curl_json -X PUT "${BASE_URL}/_search/pipeline/hybrid_rag_pipeline" \
+          -o "${ARYN_STATUSDIR}/curl.create_rag_pipeline" \
+          --data @- <<END || die "Error registering hybrid RAG pipeline"
+{
+  "phase_results_processors": [
+    {
+      "normalization-processor": {
+        "normalization": {
+          "technique": "min_max"
+        },
+        "combination": {
+          "technique": "arithmetic_mean",
+          "parameters": {
+            "weights": [0.111, 0.889]
```

Comment:
@HenryL27 is this what we use everywhere?

Reply:
this is. Approximation of 1:8 (what that blog post found was best) bc they decided weights must sum to one. I've tried a couple of other weightings, but this has performed best on our benchmark so far.